plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java

```diff
@@ -28,6 +51,99 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
 	
 	@Override
 	protected void registerOtherEditorHandlers() {
+		setupOpenType();
+	}
+	
+	private void setupOpenType() {
+		evaluateGlobalSourceStructure();
+		listenToUpdatesOfGlobalSourceStructure();
+	}
+	
+	private void evaluateGlobalSourceStructure() {
+		new Job("Searching types") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					ResourcesPlugin.getWorkspace().getRoot().accept(RustEditorActionContributor.this::addResource);
+				} catch(Exception e) {
+					AbstractLangCore.log().logError("Global source structure could not be determined", e);
+				}
+				return Status.OK_STATUS;
+			}
+		}.schedule();
+	}
+	
+	private boolean addResource(IResource resource) {
+		convertToRustFileLocation(resource).ifPresent(RustEditorActionContributor::fileTouched);
+		return true;
+	}
+	
+	private void listenToUpdatesOfGlobalSourceStructure() {
+		ResourcesPlugin.getWorkspace().addResourceChangeListener(this::processEvent, IResourceChangeEvent.POST_CHANGE);
 	}
 	
+	private void processEvent(IResourceChangeEvent event) {
+		new Job("Updating types") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					event.getDelta().accept(RustEditorActionContributor.this::applyResourceDelta);
+				} catch(Exception e) {
+					AbstractLangCore.log().logError("Global source structure could not be determined", e);
+				}
+				return Status.OK_STATUS;
+			}
+		}.schedule();
+	}
+	
+	private boolean applyResourceDelta(IResourceDelta delta) {
+		convertToRustFileLocation(delta.getResource()).ifPresent(location -> {
+			switch(delta.getKind()) {
+			case IResourceDelta.ADDED:
+			case IResourceDelta.CHANGED:
+				fileTouched(location);
+				break;
+			case IResourceDelta.REMOVED:
+				fileRemoved(location);
+				break;
+			}

```

Comment:
Note, don't use an Eclipse Job, as there is no point to that. Eclipse Jobs are good for user operations/tasks, that the user itself can see in the Progress view, and optionally cancel, but this is an internal task. Use a Java Executor instead and submit a task there. See `melnorme.lang.ide.core.engine.AbstractAgentManager` for example...